internal/parser/translate.go

```diff
@@ -327,84 +327,155 @@ func fromIngressV1(log logrus.FieldLogger, ingressList []*networkingv1.Ingress)
 
 // fromHTTPRoutes processes all the HTTPRoute objects present in the cache and translates
 // them to Kong Gateway configurations.
-//
-// TODO: this function is a WIP and is only implemented at POC level currently. The code
-//       is not reachable unless the Gateway feature gate has been enabled and the relevant
-//       milestone for tracking completion is:
-//
-//         https://github.com/Kong/kubernetes-ingress-controller/issues/2079
-//
-//       you may see several TODO's scattered throughout this function, they are all covered by #2079
-func fromHTTPRoutes(_ logrus.FieldLogger, httpRouteList []*gatewayv1alpha2.HTTPRoute) ingressRules { // TODO: use log
+func fromHTTPRoutes(log logrus.FieldLogger, httpRouteList []*gatewayv1alpha2.HTTPRoute) ingressRules {
 	result := newIngressRules()
 
 	for _, httproute := range httpRouteList {
 		// first we grab the spec and gather some metdata about the object
-		spec := httproute.Spec
 		objectInfo := util.FromK8sObject(httproute)
+		spec := httproute.Spec
 
-		// next we need to gather all the matching information e.g. paths, methods, e.t.c.
-		methods := make([]*string, 0) // TODO: implement inclusive path/method matches
-		paths := make([]*string, 0)
+		// gather the hostnames that will be used (globally) for route matching
+		hostnames := make([]*string, 0, len(spec.Hostnames))
+		for _, hostname := range spec.Hostnames {
+			hostnames = append(hostnames, kong.String(string(hostname)))
+		}
+
+		// each rule may represent a different set of backend services that will be accepting
+		// traffic, so we make separate routes and Kong services for every present rule.
 		for _, rule := range spec.Rules {
-			for _, match := range rule.Matches {
-				paths = append(paths, match.Path.Value)
+			// the HTTPRoute specification upstream specifically defines matches as
+			// independent (e.g. each match is an OR with other matches, not an AND).
+			// Therefore we treat each match rule as a separate Kong Route, so we iterate through
+			// all matches to determine all the routes that will be needed for the services.
+			routes := make([]kongstate.Route, 0, len(rule.Matches))
+			for matchNumber, match := range rule.Matches {
+				// determine the name of the route, identify it as a route that belongs
+				// to a Kubernetes HTTPRoute object.
+				routeName := kong.String(fmt.Sprintf(
+					"httproute.%s.%s.%d",
+					httproute.Namespace,
+					httproute.Name,
+					matchNumber,
```

Comment:
see #1005. Is there any criterion we could use to make matches immune to reordering?

Reply:
It's a good thought. At first I considered sorting and I also considered something like a base16 encode of object contents. I feel it's an interesting enough problem to merit it's own PR, so I've added a TODO in 56f4cc6f578672a6013e32f0b0cf1c548d9a67b5, and also added the checklist item `- [ ] add deterministic unique route identifier for HTTPRoute match translations` in https://github.com/Kong/kubernetes-ingress-controller/issues/2079 to resolve this. Let me know if that sounds good to you or if you'd like to see it handled sooner.

Reply:
sgtm :+1: